OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs

```diff
@@ -14,93 +14,80 @@
 
 namespace OpenRA.Mods.Common.Traits.Render
 {
-	public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>, Requires<ArmamentInfo>, Requires<AttackBaseInfo>
+	public class WithAttackAnimationInfo : UpgradableTraitInfo, Requires<WithSpriteBodyInfo>, Requires<ArmamentInfo>, Requires<AttackBaseInfo>
 	{
 		[Desc("Armament name")]
 		public readonly string Armament = "primary";
 
 		[Desc("Displayed while attacking.")]
 		[SequenceReference] public readonly string AttackSequence = null;
 
-		[Desc("Displayed while targeting.")]
-		[SequenceReference] public readonly string AimSequence = null;
-
-		[Desc("Shown while reloading.")]
-		[SequenceReference(null, true)] public readonly string ReloadPrefix = null;
-
 		[Desc("Delay in ticks before animation starts, either relative to attack preparation or attack.")]
 		public readonly int Delay = 0;
 
 		[Desc("Should the animation be delayed relative to preparation or actual attack?")]
 		public readonly AttackDelayType DelayRelativeTo = AttackDelayType.Preparation;
 
-		public object Create(ActorInitializer init) { return new WithAttackAnimation(init, this); }
+		public override object Create(ActorInitializer init) { return new WithAttackAnimation(init, this); }
 	}
 
-	public class WithAttackAnimation : ITick, INotifyAttack
+	public class WithAttackAnimation : UpgradableTrait<WithAttackAnimationInfo>, ITick, INotifyAttack
 	{
-		readonly WithAttackAnimationInfo info;
-		readonly AttackBase attack;
 		readonly Armament armament;
-		readonly WithSpriteBody wsb;
+		readonly WithSpriteBody[] wsbs;
 
 		int tick;
 
 		public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+			: base(info)
 		{
-			this.info = info;
-			attack = init.Self.Trait<AttackBase>();
 			armament = init.Self.TraitsImplementing<Armament>()
 				.Single(a => a.Info.Name == info.Armament);
-			wsb = init.Self.Trait<WithSpriteBody>();
+			wsbs = init.Self.TraitsImplementing<WithSpriteBody>().ToArray();
 		}
 
 		void PlayAttackAnimation(Actor self)
 		{
-			if (!string.IsNullOrEmpty(info.AttackSequence))
-				wsb.PlayCustomAnimation(self, info.AttackSequence, () => wsb.CancelCustomAnimation(self));
+			foreach (var wsb in wsbs)
+				if (!wsb.IsTraitDisabled && !string.IsNullOrEmpty(Info.AttackSequence))
+					wsb.PlayCustomAnimation(self, Info.AttackSequence, () => wsb.CancelCustomAnimation(self));
```

Comment:
Thinking aloud, might replacing this `IsTraitDisabled` check with an identifier/`Name` lookup (kind of like what we do with armaments already) help, or does it depend on the body itself rather than the modifying trait?